servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/MpmcSequentialRunQueue.java

```diff
@@ -0,0 +1,185 @@
+            }
+        }
+    }
+
+    private boolean offerUpdateStaleTail(final Node node, final Node oldTail) {
+        // tail is stale so attempt to iterate through the linked list and get the current tail reference.
+        Node currentTail = oldTail.iterateToTail();
+
+        // Best effort check to see if the node is popped, then we attempt to directly insert node.
+        if (currentTail.isPopped()) {
+            if (tailUpdater.compareAndSet(this, oldTail, node)) {
+                node.run();
+                return true;
+            }
+        } else {
+            // It is possible the currentTail has been popped after our best-effort check above. That is OK because
+            // the offer method accounts for stale tail replacement.
+            tailUpdater.compareAndSet(this, oldTail, currentTail);
+        }
+        // Best effort to update/clear the tail, and failure is OK because:
+        // 1. Another thread is in offer and already patched up the tail pointer and we will read the new
+        // tail on the next loop iteration.
+        return false;
+    }
+
+    /**
+     * Pop {@code node} from this queue. This will invoke {@link Node#run()} on the next head of the queue
+     * (if one exists).
+     * <p>
+     * Typically invoked by a single thread but maybe invoked by multiple threads, and maybe invoked multiple times.
+     * Re-entry from {@link Node#run()} from another thread is permitted.
+     * @param head A {@link Node} which has been passed to {@link #offer(Node)}, and whose {@link Node#run()} has been
+     * invoked by this queue.
+     */
+    void poll(final Node head) {
+        // This method maybe called multiple times on the same node, in which case next will be EMPTY_NODE and the run
+        // method will be a noop.
+        Node next = head.pop();
+        if (next != null) {
+            next.run();
+        } else {
+            tailUpdater.compareAndSet(this, head, null);
+            // Best effort to clear the tail, and failure is OK because:
+            // 1. Another thread appended this head, but has not yet updated the tail. In this case the tail will be
+            // stale (e.g. pointing to head node that has already been processed) and corrected by future inserts.
+        }
+    }
+
+    /**
+     * Single linked node in the queue.
+     * <p>
+     * Instances are single use only! No re-use or sharing between queue instances!
+     */
+    abstract static class Node {
+        /**
+         * the EMPTY_NODE's next must point to itself so that attempts to {@link #append(Node)} on it will fail,
+         * and also be detected as {@link #isPopped()} and not prevent any new nodes from being
+         * {@link #offer(Node) offered}.
+         */
+        private static final Node EMPTY_NODE = new Node(false) {
+            @Override
+            void run() {
+            }
+        };
+        private static final AtomicReferenceFieldUpdater<Node, Node> nextUpdater =
```

Comment:
nit: you could just have `Node` extend `AtomicReference<Node>` and remove the whole updater dance. This will also remove some overhead in terms of access checks.